kernel/sched.c

```diff
@@ -976,8 +976,17 @@ void *z_get_next_switch_handle(void *interrupted)
 			if (z_is_thread_queued(old_thread)) {
 				runq_add(old_thread);
 			}
+		} else {
+			/* Re-set the switch_handle here if we don't have to
+			 * context switch later (because old_thread ==
+			 * new_thread).
+			 *
+			 * In case of context switching the switch_handle will
+			 * be set later after the context is being swapped out
+			 * to unblock the wait_for_switch().
+			 */
+			old_thread->switch_handle = interrupted;
```

Comment:
I'm having trouble understanding how this works.  How exactly does this not break CPU migration?

If old_thread != new_thread (i.e. we're about to switch), then this won't set old_thread's switch handle to the state saved by ISR entry, and thus it's going to be left NULL.  So whichever CPU picks that up in the future won't have saved context to switch into, right?

And if old_thread == new_thread, then it maintains the existing behavior modulo a branch and can't change any behavior except for maybe shrinking a race window.

I totally buy this is a race, I just don't think it's here.